### PR TITLE
Adding addional HTTP header options

### PR DESCRIPTION
**Pull request by tmughal**
This commit adds support for bearer authentication during the WebSocket handshake request.

### CODE REVIEW

**Comment by tmughal** [comment]
@youviewtv/ios Please review

File: SocketRocket/SRWebSocket.m [hunk 1]
```diff
@@ -373,6 +373,25 @@ - (void)_HTTPHeadersDidFinish;
...
+        {
+            // handle an aunathorised request to upgrade
+            if ([self.authenticationDelegate respondsToSelector:@selector(webSocket:bearerAuthenticationDidFailWithRetryHandler:)])
+            {
+                [self.authenticationDelegate webSocket:self bearerAuthenticationDidFailWithRetryHandler:^(NSString *newBearerToken) {
+                    if (newBearerToken)
+                    {
+                        dispatch_async(_workQueue, ^{
+                            self.authenticationBearerToken = newBearerToken;
+                            [self didConnect];
```
Comment (wandeep) [inline SocketRocket/SRWebSocket.m]: Use weakSelf

**Review by wandeep — APPROVED** [review]
+1 With minor comment for consideration - also any tests required for this change?

File: SocketRocket/SRWebSocket.m [hunk 2]
```diff
@@ -373,6 +373,25 @@ - (void)_HTTPHeadersDidFinish;
 {
     NSInteger responseCode = CFHTTPMessageGetResponseStatusCode(_receivedHTTPHeaders);
     if (responseCode >= 400) {
+        
+        if (responseCode == 401)
```
Comment (neildavis) [inline SocketRocket/SRWebSocket.m]: You should also check for a "WWW-Authenticate" HTTP header in the response identifying the "Bearer" scheme before relying on the authentication delegate to resolve. There are other reasons for 401 responses

File: SocketRocket/SRWebSocket.m (hunk 1 above)
Comment (neildavis) [inline SocketRocket/SRWebSocket.m]: How do you resolve the infinite loop case when the delegate provides a token but you still receive a 401? If the token provided by the delegate does not resolve the authentication then the 401 error should be propogated to the caller. This is analagous to the HTTP middleware protocol 'fail fast' approach

**Review by neildavis — CHANGES_REQUESTED** [review]

File: SocketRocket/SRWebSocket.m (hunk 2 above)
Comment (tmughal) [reply SocketRocket/SRWebSocket.m]: Funny enough, I was just thinking this last night. Made the change first thing this morning :)

File: SocketRocket/SRWebSocket.m (hunk 1 above)
Comment (tmughal) [reply SocketRocket/SRWebSocket.m]: Shouldn't really need to use weakSelf as the block doesn't have a copy reference to it.

**Comment by tmughal** [comment]
@youviewtv/ios Please review following changes.

This now allows additional headers to be set that will be used as part of the protocol handshake. Additionally, the error passed through the delegate will contain received headers as part of the user info when applicable.

File: SocketRocket/SRWebSocket.h [hunk 3]
```diff
@@ -122,6 +127,11 @@ extern NSString *const SRHTTPResponseErrorKey;
  */
 @property (nonatomic, assign, readonly) BOOL allowsUntrustedSSLCertificates;
 
+/**
+ Additional HTTP request headers to append for the protocol handshake.
+ */
+@property (nonatomic, strong) NSDictionary<NSString *, NSString *> *additionalHttpHeaders;
```
Comment (neildavis) [inline SocketRocket/SRWebSocket.h]: HTTP is capitalized everywhere else in the API

**Review by neildavis — APPROVED** [review]
+1 with comment

**Review by faisalyv — APPROVED** [review]
+1

**Pull request closed without merging**